rustls/src/server/verifier_builder.rs

```diff
@@ -0,0 +1,268 @@
+use crate::verify::{AnonymousClientPolicy, ClientCertVerifier, WebPkiClientVerifier};
+use crate::{CertRevocationListError, RootCertStore};
+
+/// A builder for configuring a `webpki` client certificate verifier.
+///
+/// For more information, see the [`WebPkiClientVerifier`] documentation.
+#[derive(Debug, Clone)]
+pub struct ClientCertVerifierBuilder {
+    roots: RootCertStore,
+    crls: Vec<UnparsedCertRevocationList>,
+    anon_policy: AnonymousClientPolicy,
+}
+
+impl ClientCertVerifierBuilder {
+    pub(crate) fn new(roots: RootCertStore) -> Self {
+        Self {
+            roots,
+            crls: Vec::new(),
+            anon_policy: AnonymousClientPolicy::Forbid,
+        }
+    }
+
+    /// Add additional `roots` to use to verify client certificates.
+    ///
+    /// All clients must provide a client certificate unless you have allowed unauthenticated
+    /// clients by calling [allow_unauthenticated()][ClientCertVerifierBuilder::allow_unauthenticated]
+    /// on the builder.
+    pub fn with_roots(mut self, roots: RootCertStore) -> Self {
+        self.roots
+            .add_server_trust_anchors(roots.roots.into_iter());
+        self
+    }
+
+    /// Verify the revocation state of presented client certificates against the provided
+    /// certificate revocation lists (CRLs). Calling `with_crls` multiple times appends the
+    /// given CRLs to the existing collection.
+    pub fn with_crls(mut self, crls: impl IntoIterator<Item = UnparsedCertRevocationList>) -> Self {
+        self.crls.extend(crls);
+        self
+    }
+
+    /// Allow unauthenticated clients to connect.
+    ///
+    /// Clients that offer a client certificate issued by a trusted root, and clients that offer no
+    /// client certificate will be allowed to connect.
+    pub fn allow_unauthenticated(mut self) -> Self {
+        self.anon_policy = AnonymousClientPolicy::Allow;
+        self
+    }
+
+    /// Build a client certificate verifier. The built verifier will be used for the server to offer
+    /// client certificate authentication, to control how offered client certificates are validated,
+    /// and to determine what to do with anonymous clients that do not respond to the client
+    /// certificate authentication offer with a client certificate.
+    ///
+    /// Once built, the provided `Arc<dyn ClientCertVerifier>` can be used with a Rustls
+    /// [crate::server::ServerConfig] to configure client certificate validation using
+    /// [`with_client_cert_verifier`][crate::ConfigBuilder<ClientConfig, WantsVerifier>::with_client_cert_verifier].
+    ///
+    /// # Errors
+    /// This function will return a `ClientCertVerifierBuilderError` if:
+    /// 1. No trust anchors have been provided.
+    /// 2. DER encoded CRLs have been provided that can not be parsed successfully.
+    pub fn build(self) -> Result<Arc<dyn ClientCertVerifier>, ClientCertVerifierBuilderError> {
+        match self.roots.is_empty() {
```

Comment:
Similar to the webpki thing from earlier, maybe just make this an early returning `if`?

Reply:
Done.